src/util/fifo.h

```diff
@@ -1,69 +1,116 @@
 #pragma once
 
-#include "pa_ringbuffer.h"
+#include <cstdint>
 
 #include "util/class.h"
 #include "util/math.h"
 
+using ring_buffer_size_t = uint32_t;
+
 template <class DataType>
 class FIFO {
   public:
-    explicit FIFO(int size)
-            : m_data(roundUpToPowerOf2(size)) {
-        // If we can't represent the next higher power of 2 then bail.
-        if (m_data.size() == 0) {
-            return;
-        }
-        PaUtil_InitializeRingBuffer(&m_ringBuffer,
-                static_cast<ring_buffer_size_t>(sizeof(DataType)),
-                static_cast<ring_buffer_size_t>(m_data.size()),
-                m_data.data());
+    explicit FIFO(uint32_t size)
+            : m_size{roundUpToPowerOf2(size)},
+              m_mask{m_size - 1},
+              m_data(m_size),
+              m_writeIndex{0},
+              m_readIndex{0} {
     }
-    virtual ~FIFO() {
+    uint32_t readAvailable() const {
+        const uint32_t readIndex = m_readIndex.load(std::memory_order_relaxed);
+        const uint32_t writeIndex = m_writeIndex.load(std::memory_order_acquire);
+        return writeIndex - readIndex;
```

Comment:
Can you add a comment about the case if a read or write happens between the wo atomic assesses?

Reply:
Cam you please also explain the use of the memory barriers? Is it correct?

Reply:
Yes, I can write some comments about the atomics and the memory barriers.

Reply:
comments added. please review and resolve.

Reply:
I did not see comment about the memory barriers and what happens if read and write happens in between reading the two atomics. Is it not yet pushed?

Reply:
It is the last comment, after the declaration of the two atomics.